pinot-common/src/main/java/org/apache/pinot/common/utils/ServiceStartableUtils.java

```diff
@@ -36,56 +39,91 @@ private ServiceStartableUtils() {
-        .setSessionTimeout(zkClientSessionConfig)
-        .build();
+    ZkClient zkClient = new ZkClient.Builder().setZkSerializer(new ZNRecordSerializer()).setZkServer(zkAddress)
+        .setConnectionTimeout(zkClientConnectionTimeoutMs).setSessionTimeout(zkClientSessionConfig).build();
     zkClient.waitUntilConnected(zkClientConnectionTimeoutMs, TimeUnit.MILLISECONDS);
+    return zkClient;
+  }
 
-    try {
-      ZNRecord clusterConfigZNRecord =
-          zkClient.readData(String.format(CLUSTER_CONFIG_ZK_PATH_TEMPLATE, clusterName, clusterName), true);
-      if (clusterConfigZNRecord == null) {
-        LOGGER.warn("Failed to find cluster config for cluster: {}, skipping applying cluster config", clusterName);
-        return;
-      }
+  /**
+   * Overrides the instance config with the tenant configs if the tenant is tagged on the instance.
+   */
+  public static void applyTenantConfigs(String instanceId, ZkClient zkClient, String clusterName,
+      PinotConfiguration instanceConfig) {
+    ZNRecord instanceConfigZNRecord =
+        zkClient.readData(String.format(INSTANCE_CONFIG_ZK_PATH_TEMPLATE, clusterName, instanceId), true);
+    if (instanceConfigZNRecord == null) {
+      LOGGER.warn("Failed to find instance config for instance: {}, skipping overriding tenant configs", instanceId);
+      return;
+    }
+    InstanceConfig instanceZKConfig = new InstanceConfig(instanceConfigZNRecord);
+    Set<String> instanceTags = instanceZKConfig.getTags().stream()
+        .map(PinotConfiguration::relaxPropertyName)
+        .collect(Collectors.toSet());
 
-      Map<String, String> clusterConfigs = clusterConfigZNRecord.getSimpleFields();
-      String instanceConfigKeyPrefix =
-          String.format(PINOT_INSTANCE_CONFIG_KEY_PREFIX_TEMPLATE, serviceRole.name().toLowerCase());
-      for (Map.Entry<String, String> entry : clusterConfigs.entrySet()) {
-        String key = entry.getKey();
-        String value = entry.getValue();
-        if (key.startsWith(PINOT_ALL_CONFIG_KEY_PREFIX)) {
-          String instanceConfigKey = instanceConfigKeyPrefix + key.substring(PINOT_ALL_CONFIG_KEY_PREFIX.length());
-          addConfigIfNotExists(instanceConfig, instanceConfigKey, value);
-        } else {
-          // TODO: Currently it puts all keys to the instance config. Consider standardizing instance config keys and
-          //       only put keys with the instance config key prefix.
-          addConfigIfNotExists(instanceConfig, key, value);
+    for (String key : instanceConfig.getKeys()) {
+      if (key.startsWith(PINOT_TAG_LEVEL_CONFIG_KEY_PREFIX)) {
+        String instanceConfigKey = key.substring(PINOT_TAG_LEVEL_CONFIG_KEY_PREFIX.length());
+        String tag = instanceConfigKey.substring(0, instanceConfigKey.indexOf('.'));
+        String tagKey = instanceConfigKey.substring(tag.length() + 1);
+        if (instanceTags.contains(tag)) {
+          instanceConfig.setProperty(tagKey, instanceConfig.getProperty(key));
```

Comment:
Do we need the complete key here? Or we just need the part after `pinot.<type>`?
E.g. Do we want `pinot.tag.myTenant_OFFLINE.pinot.server.query.executor.abc` or `pinot.tag.myTenant_OFFLINE.query.executor.abc`?